pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -797,32 +836,57 @@ func (n *NetworkPolicyController) deleteNetworkPolicy(old interface{}) {
 		n.enqueueInternalGroup(group)
 	}
+
+	// If this Service doesn't have a selector, there won't be an Endpoints ADD event
+	// triggered by this Service ADD event. So find all ACNPs that are impacted by this
+	// Service and reprocess them here. Otherwise, an Endpoints ADD event will
+	// reprocess those ACNPs.
+	if service.Spec.Selector == nil {
+		affectedACNPsByService := n.filterACNPsByService(service.Namespace, service.Name)
+		for _, acnp := range affectedACNPsByService {
+			n.reprocessACNP(acnp)
+		}
+	}
 }
 
-// updatePod retrieves all internal Groups which refers to this Service
-// and enqueues the group keys for further processing.
+// updateService retrieves all internal Groups which refers to this Service
+// and enqueues the group keys for further processing. And reprocess the
+// ClusterNetworkPolicy with `toService` referred to this Service.
 func (n *NetworkPolicyController) updateService(oldObj, curObj interface{}) {
 	defer n.heartbeat("updateService")
 	oldService := oldObj.(*v1.Service)
 	curService := curObj.(*v1.Service)
-	klog.V(2).Infof("Processing Service %s/%s UPDATE event, selectors: %v", curService.Namespace, curService.Name, curService.Spec.Selector)
-	// No need to trigger processing of groups if there is no change in the Service selectors.
-	if reflect.DeepEqual(oldService.Spec.Selector, curService.Spec.Selector) {
-		klog.V(4).Infof("No change in Service %s/%s. Skipping group evaluation.", curService.Namespace, curService.Name)
+	klog.V(2).InfoS("Processing Service UPDATE event", "serviceNamespace", curService.Namespace, "serviceName", curService.Name)
+	// Only trigger processing of groups if there are changes in the Service selectors.
+	if !reflect.DeepEqual(oldService.Spec.Selector, curService.Spec.Selector) {
+		// Find all internal Group keys which refers to this Service.
+		groupKeySet := n.filterInternalGroupsForService(curService)
+		// Enqueue internal groups to its queue for group processing.
+		for group := range groupKeySet {
+			n.enqueueInternalGroup(group)
+		}
+		// If the Selector of this Service has been updated, an Endpoints UPDATE event will
+		// reprocess all ACNPs that are impacted by this Service. No need to reprocess them
+		// here.
 		return
 	}
-	// Find all internal Group keys which refers to this Service.
-	groupKeySet := n.filterInternalGroupsForService(curService)
-	// Enqueue internal groups to its queue for group processing.
-	for group := range groupKeySet {
-		n.enqueueInternalGroup(group)
+
+	// If this Service doesn't have a Selector or both Selector and Ports haven't been
+	// updated, then there won't be an Endpoints UPDATE event triggered by this Service
+	// UPDATE event. So find all ACNPs that are impacted by this Service and reprocess
+	// them here. Otherwise, an Endpoints UPDATE event will reprocess those ACNPs.
+	if curService.Spec.Selector == nil || (curService.Spec.Selector != nil && reflect.DeepEqual(oldService.Spec.Ports, curService.Spec.Ports)) {
```

Comment:
why do the ACNPs need to be reprocessed when the ports have not been updated?

Reply:
I'm trying to cover all cases that Service has an update while the Endpoints is not. Like ServiceType.

Reply:
I kind of have the same question as Abhishek. I don't understand why the first clause of the test is `curService.Spec.Selector == nil` and not `(curService.Spec.Selector == nil && oldService.Spec.Selector != nil)`

Reply:
Because this case: (curService.Spec.Selector == nil && oldService.Spec.Selector != nil)  has been checked above.

Reply:
Oh I realized that if the `curService.Spec.Selector == nil` the update on targetPort won't be reflected on the Endpoints. No matter the targetPort change or not, traffic will always be forward to the Port of Endpoints.

Reply:
So it seems that in all cases that Service has an update while the Endpoints is not, only ServiceType update should trigger the reprocess. What do you think?